test/functional/test_framework/key.py

```diff
@@ -431,6 +471,41 @@ def sign_schnorr(self, msg):
         e = int.from_bytes(hashlib.sha256(R[0].to_bytes(32, 'big') + self.get_pubkey().get_bytes() + msg).digest(), 'big') % SECP256K1_ORDER
         return R[0].to_bytes(32, 'big') + ((k + e*self.secret) % SECP256K1_ORDER).to_bytes(32, 'big')
 
+    def sign_schnorr_with_nonce(self, msg, k_key):
+        """Construct a bip-schnorr with this key and a predetermined nonce."""
+        assert(self.valid)
+        assert(self.compressed)
+        assert(len(msg) == 32)
+        if k_key == None:
+            kp = int.from_bytes(hashlib.sha256(self.get_bytes() + msg).digest(), 'big') % SECP256K1_ORDER
+        else:
+            kp = int.from_bytes(k_key.get_bytes(), 'big') % SECP256K1_ORDER
+        assert(kp != 0)
+        R = SECP256K1.affine(SECP256K1.mul([(SECP256K1_G, kp)]))
+        if jacobi_symbol(R[1], SECP256K1_FIELD_SIZE) == 1:
+            k = kp
+        else:
+            k = SECP256K1_ORDER - kp
+        e = int.from_bytes(hashlib.sha256(R[0].to_bytes(32, 'big') + self.get_pubkey().get_bytes() + msg).digest(), 'big') % SECP256K1_ORDER
+        return R[0].to_bytes(32, 'big') + ((k + e*self.secret) % SECP256K1_ORDER).to_bytes(32, 'big')
+
+    def sign_musig(self, k_key, R_negated, R_musig, P_musig, msg):
```

Comment:
Is it possible to remove the `R_negated` parameter here, and just have the individual signers update their `r` values to negate them when necessary?